src/webgpu/api/validation/createRenderPipeline.spec.ts

```diff
@@ -248,6 +251,145 @@ g.test('color_formats_must_be_renderable')
+    t.doCreateRenderPipelineTest(isAsync, !depthTestEnabled || info.depth, descriptor);
+  });
+
+g.test('depth_stencil_state,depth_aspect,depth_write')
+  .desc(
+    `Depth aspect must be contained in the format if depth write is enabled in depthStencilState.`
+  )
+  .params(u =>
+    u
+      .combine('isAsync', [false, true])
+      .combine('format', kDepthStencilFormats)
+      .combine('depthWriteEnabled', [false, true])
+  )
+  .fn(async t => {
+    const { isAsync, format, depthWriteEnabled } = t.params;
+    const info = kTextureFormatInfo[format];
+    await t.selectDeviceOrSkipTestCase(info.feature);
+
+    const descriptor = t.getDescriptor({
+      depthStencil: { format, depthWriteEnabled },
+    });
+    t.doCreateRenderPipelineTest(isAsync, !depthWriteEnabled || info.depth, descriptor);
+  });
+
+g.test('depth_stencil_state,stencil_aspect,stencil_test')
+  .desc(
+    `Stencil aspect must be contained in the format if stencil test is enabled in depthStencilState.`
+  )
+  .params(u =>
+    u
+      .combine('isAsync', [false, true])
+      .combine('format', kDepthStencilFormats)
+      .combine('face', ['front', 'back'])
+      .combine('compare', [undefined, ...kCompareFunctions])
+  )
+  .fn(async t => {
+    const { isAsync, format, face, compare } = t.params;
+    const info = kTextureFormatInfo[format];
+    await t.selectDeviceOrSkipTestCase(info.feature);
+
+    let descriptor: GPURenderPipelineDescriptor;
+    if (face === 'front') {
+      descriptor = t.getDescriptor({ depthStencil: { format, stencilFront: { compare } } });
+    } else {
+      descriptor = t.getDescriptor({ depthStencil: { format, stencilBack: { compare } } });
+    }
+
+    const stencilTestEnabled = compare !== undefined && compare !== 'always';
+    t.doCreateRenderPipelineTest(isAsync, !stencilTestEnabled || info.stencil, descriptor);
+  });
+
+g.test('depth_stencil_state,stencil_aspect,stencil_write')
+  .desc(
+    `Stencil aspect must be contained in the format if stencil write is enabled in depthStencilState.`
+  )
+  .params(u =>
+    u
+      .combine('isAsync', [false, true])
+      .combine('format', kDepthStencilFormats)
+      .combine('faceAndOpType', [
+        'frontFailOp',
+        'frontDepthFailOp',
+        'frontPassOp',
+        'backFailOp',
+        'backDepthFailOp',
+        'backPassOp',
+      ])
+      .combine('op', [undefined, ...kStencilOperations])
+  )
+  .fn(async t => {
+    const { isAsync, format, faceAndOpType, op } = t.params;
+    const info = kTextureFormatInfo[format];
+    await t.selectDeviceOrSkipTestCase(info.feature);
+
+    let descriptor: GPURenderPipelineDescriptor;
+    switch (faceAndOpType) {
+      case 'frontFailOp':
+        descriptor = t.getDescriptor({ depthStencil: { format, stencilFront: { failOp: op } } });
+        break;
+      case 'frontDepthFailOp':
+        descriptor = t.getDescriptor({
+          depthStencil: { format, stencilFront: { depthFailOp: op } },
+        });
```

Comment:
uNit: could move `t.getDescriptor` out of the switch just so that none of the lines wrap.

Reply:
Good suggestion!